server/parser.go

```diff
@@ -160,7 +159,24 @@ func (c *client) parse(buf []byte) error {
 			c.op = b
 			if b != 'C' && b != 'c' {
 				if authSet {
-					goto authErr
+					if s == nil {
+						goto authErr
+					}
+					var ok bool
+					// Check here for NoAuthUser. If this is set allow non CONNECT protos as our first.
+					// E.g. telnet proto demos.
+					if noAuthUser := s.getOpts().NoAuthUser; noAuthUser != _EMPTY_ {
+						s.mu.Lock()
+						user, exists := s.users[noAuthUser]
+						s.mu.Unlock()
+						if exists {
+							c.RegisterUser(user)
+							ok = true
```

Comment:
For clients that do never send CONNECT, we will be in this code for every protocol message, since we don't clear the `authSet`. This will cause multiple registrations, no?

I am not convinced that we should allow clients without a CONNECT. The no_auth_user simply means that if no user is specified in the CONNECT, use the one configured, but it does not mean that clients should not send a CONNECT.

Also, your new test checks that the client gets disconnected anyway after the auth timeout. So what is the point of all that? Why let a client do pub/sub for the time between it TCP connects and the auth timeout?

Reply:
For the protocol demo and doing telnet. Agree I should clear the authSet though. Will update.

Reply:
I feel no_auth setting should behave same as simple server. This was only thing it was very different on.